### DISCUSSION TITLE
Macro expansion during JIT compilation?

### DISCUSSION THREAD

**Discussion by NamjaeChoi** [CEED/libCEED#1697] [Q&A] [answered]
Hello,

I'm wondering if JIT compilation can properly expand macros. I guess it is somehow treating macros since `CEED_QFUNCTION` is also a macro, but I'm not sure if you have a special rule to specifically handle it or it is being handled by the compiler. The reason for this question is because we are defining QFunctions using macros like:

```
#define CEED_KERNEL(name, parent)                                                                  \
  class name final : public parent<name>                                                           \
  {                                                                                                \
  public:                                                                                          \
    static CEEDQFunctionPointer _qf;                                                               \
    auto getQFunction() { return _qf; }                                                            \
    std::string getQFunctionPath() { return std::string(__FILE__) + ":" + #name + "_qf"; }
#define CEED_KERNEL_END(name)                                                                      \
  }                                                                                                \
  ;                                                                                                \
                                                                                                   \
  CEED_QFUNCTION(name##_qf)                                                                        \
  (void * ctx, const CeedInt Q, const CeedScalar * const * in, CeedScalar * const * out)           \
  {                                                                                                \
    auto kernel = static_cast<name *>(ctx);                                                        \
    kernel->computeResidualInternal(Q, in, out);                                                   \
    return 0;                                                                                      \
  }                                                                                                \
                                                                                                   \
  CEEDQFunctionPointer name::_qf = name##_qf;
```

which is used in the user code like:

```
CEED_KERNEL(CEEDDiffusionKernel, CEEDDiffusion)
public:
static InputParameters validParams();

CEEDDiffusionKernel(const InputParameters & parameters);
CEED_KERNEL_END(CEEDDiffusionKernel)
```

But this leads to a JIT error that the QFunction was not defined:
```
/data/choin/moose/libCEED/backends/cuda/ceed-cuda-compile.cpp:125 in CeedCompile_Cuda(): NVRTC_ERROR_COMPILATION
default_program(63222): error: identifier "CEEDDiffusionKernel_qf" is undefined
      CEEDDiffusionKernel_qf(ctx, 1, inputs, outputs);
```

**Comment by jeremylt** [comment]
I'll try to make a running example of this to reproduce and understand what's happening, but it will take me time to do so.

**Reply by jeremylt** [reply]
If you aren't familiar, a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example) greatly helps library maintainers in seeing the issue you are seeing.

**Reply by NamjaeChoi** [reply]
Let me try. This might be easy to make an example.

**Comment by jeremylt** [comment]
For CUDA JiT, the `CEED_QFUNCTION` macro is defined as
```
#define CEED_QFUNCTION(name) inline __device__ int name
```
I would guess you are  getting a mixup with macro composition?

**Reply by jeremylt** [reply]
You can set the environment variable `CEED_DEBUG=1` to see the code being JiTed, though that will potentially go away in the future as a result of https://github.com/CEED/libCEED/discussions/1695

**Reply by NamjaeChoi** [reply]
Hmm that just give me the code before expanding macro.

**Reply by jeremylt** [reply]
Ah, that's right.  I'm still suspicious of chaining string manipulation macros. That's very touchy in my experience

**Reply by NamjaeChoi** [reply]
I just tried chaining macros in `ex1-volume` but seems to work fine. Let look into this more.

**Comment by NamjaeChoi** [comment, marked as answer]
Ah I think I know the reason. All our CEED source files are protected with a preprocessor, which is defined by our static compiler. But I guess JIT compiler doesn't know that? Is there a way to provide a preprocessor to the JIT compiler?

**Reply by jeremylt** [reply]
Do you mean there is some `#ifdef MOOSE_IS_USING_LIBCEED` declared?

**Reply by NamjaeChoi** [reply]
Exactly, in order to prevent them from being compiled when libCEED is disabled.

**Reply by jedbrown** [reply]
Would it be painful to reverse it, `MOOSE_IGNORE_LIBCEED`? I'm not opposed to adding support for macro definitions (I assume all our online compilers will accept `-DMOOSE_USING_LIBCEED=1`). An alternative would be to include a MOOSE `config.h` in each compilation unit.

**Reply by NamjaeChoi** [reply]
Like `#ifndef MOOSE_IGNORE_LIBCEED`? It would probably be doable. Actually it's a good suggestion.

**Reply by NamjaeChoi** [reply]
Thank you, this issue was addressed, although I'm struck by another inclusion error. Now the inclusion error is coming from the compiler:
```
/data/choin/moose/libCEED/backends/cuda/ceed-cuda-compile.cpp:125 in CeedCompile_Cuda(): NVRTC_ERROR_COMPILATION
default_program(375): catastrophic error: cannot open source file "Kokkos_Core.hpp"
  #include <Kokkos_Core.hpp>
```
Jeremy's PR addressed the inclusion error of the libCEED side, but that information doesn't seem to be relayed to the backend compiler. Do you have any idea?

**Reply by jeremylt** [reply]
Note - this issue with the macro definition `MOOSE_IS_USING_LIBCEED` is a good example of why I recommend that the transitive includes of the QFunction source file you pass should contain only the absolute minimum needed for compilation and nothing extra. This is why `ceed/types.h` exists, for example.

This error
```
/data/choin/moose/libCEED/backends/cuda/ceed-cuda-compile.cpp:125 in CeedCompile_Cuda(): NVRTC_ERROR_COMPILATION
default_program(375): catastrophic error: cannot open source file "Kokkos_Core.hpp"
  #include <Kokkos_Core.hpp>
```
would indicate that `Kokkos_Core.hpp` is not in one of the directories you set with `CeedAddJitSourceRoot`.

Note - the change I made was to `ceed-cuda-compile.cpp`.

**Reply by NamjaeChoi** [reply]
The following error:
```
/data/choin/moose/libCEED/interface/ceed-jit-tools.c:138 in CeedLoadSourceToInitializedBuffer(): Couldn't open source file: /data/choin/projects/moose/framework/build/header_symlinks/Kokkos_Core.hpp
```
went away with your PR plus changing to `<header.h>` and setting the directory by `CeedAddJitSourceRoot`. If the directory was not set, I believe the prior stage wouldn't have passed.

I doubt but, would there be any potential issue if I run things in a container?

**Reply by jeremylt** [reply]
It would be so much easier for me to understand with an example of the code that I can run myself.

I'd make sure you are using the absolute latest version of my branch as I made updates to it through the day. I fixed an error with passing the directory list through an inheritance chain recently.That error message I can only interpret as the ceed not finding the file in the directories it knows about.

**Reply by NamjaeChoi** [reply]
I fully understand your difficulty and would love to make a reproduce example, but our code is very heavy. So I'm trying to debug on my front as much as possible. One thing I noticed is that two `Ceed`s are being interchangeably used in `CeedCompile_Cuda`, even though I only created one. One has all the JIT roots properly associated, while the other is not, which is causing the issue.

**Reply by jeremylt** [reply]
That sounds like the issue I fixed earlier today where inheritance was not passing the list of directories around correctly.

**Reply by NamjaeChoi** [reply]
I'm at the latest state of your branch. What is fallback `Ceed`? This is being created during `CeedOperatorApply` and passed through the chain of  `CeedOperatorCreateFallback -> CeedQFunctionCreateFallback -> CeedQFunctionCreateInterior` but doesn't seem to hold the JIT root information of the original `Ceed`. Sorry I have literally no knowledge of the internal structure of libCEED.

**Reply by jeremylt** [reply]
That is the inheritance mechanism. The inheritance parent holds the list and my changes check the list held by the parent.

If you cannot provide code, the full callstack could help. I thought Moose was open source though?

**Reply by NamjaeChoi** [reply]
The code itself is already in my fork. It's just that compiling and running MOOSE will be a pain for you.

The following is the stack trace when different `Ceed` is being used from my `Ceed` that doesn't give the proper JIT root information.

```
Thread 1 "gpu_app-dbg" hit Breakpoint 1, CeedCompile_Cuda (ceed=0x2467ab0, source=0x30e6fe0 "\n// Copyright (c) 2017-2024, Lawrence Livermore National Security, LLC and other CEED contributors.\n// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.\n//\n// SPDX-License-I"..., module=0x2625150, num_defines=0) at /data/choin/moose/libCEED/backends/cuda/ceed-cuda-compile.cpp:37
37      int CeedCompile_Cuda(Ceed ceed, const char *source, CUmodule *module, const CeedInt num_defines, ...) {
(gdb) bt
#0  CeedCompile_Cuda (ceed=0x2467ab0, 
    source=0x30e6fe0 "\n// Copyright (c) 2017-2024, Lawrence Livermore National Security, LLC and other CEED contributors.\n// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.\n//\n// SPDX-License-I"..., module=0x2625150, num_defines=0) at /data/choin/moose/libCEED/backends/cuda/ceed-cuda-compile.cpp:37
#1  0x00007fffee468f4e in CeedQFunctionBuildKernel_Cuda_ref (qf=qf@entry=0x261c0d0) at /opt/rh/gcc-toolset-12/root/usr/include/c++/12/bits/basic_string.h:233
#2  0x00007fffee45cbf8 in CeedQFunctionApply_Cuda (qf=0x261c0d0, Q=<optimized out>, U=0x29f3500, V=0x23b6c70) at /data/choin/moose/libCEED/backends/cuda-ref/ceed-cuda-ref-qfunction.c:30
#3  0x00007fffee42db7d in CeedQFunctionApply (qf=0x261c0d0, Q=400, u=0x29f3500, v=0x23b6c70) at /data/choin/moose/libCEED/interface/ceed-qfunction.c:1054
#4  0x00007fffee45c2a5 in CeedOperatorApplyAdd_Cuda (op=<optimized out>, in_vec=0x7fffee61d880 <ceed_vector_none>, out_vec=0x2012a70, request=0x7fffee61d868 <ceed_request_immediate>)
    at /data/choin/moose/libCEED/backends/cuda-ref/ceed-cuda-ref-operator.c:532
#5  0x00007fffee427e8d in CeedOperatorApplyAdd (op=0x17ed540, in=in@entry=0x7fffee61d880 <ceed_vector_none>, out=out@entry=0x2012a70, 
    request=request@entry=0x7fffee61d868 <ceed_request_immediate>) at /data/choin/moose/libCEED/interface/ceed-operator.c:2183
#6  0x00007fffee465995 in CeedOperatorApplyAdd_Cuda_gen (op=0x256c1f0, input_vec=0x7fffee61d880 <ceed_vector_none>, output_vec=0x2012a70, request=0x7fffee61d868 <ceed_request_immediate>)
    at /data/choin/moose/libCEED/backends/cuda-gen/ceed-cuda-gen-operator.c:126
#7  0x00007fffee427e8d in CeedOperatorApplyAdd (op=0x256c1f0, in=0x7fffee61d880 <ceed_vector_none>, out=0x2012a70, request=0x7fffee61d868 <ceed_request_immediate>)
    at /data/choin/moose/libCEED/interface/ceed-operator.c:2183
```

**Reply by jeremylt** [reply]
That logic path given by that call stack is tested by CI in my branch and is working there.

Which brings me back to the theory the JiT include dirs added aren't correct?

You can bypass the inheritance by using `/gpu/cuda/ref` instead of `gen` but both are tested in CI.

**Reply by NamjaeChoi** [reply]
I just tried the following (using `op_fallback_parent` instead of `parent`) for testing and now JIT source information is correct.

```
int CeedGetJitSourceRoots(Ceed ceed, CeedInt *num_source_roots, const char ***jit_source_roots) {
  Ceed ceed_parent;

  CeedCall(CeedGetParent(ceed, &ceed_parent));
  if (ceed->op_fallback_parent) ceed_parent = ceed->op_fallback_parent;
```

**Reply by jeremylt** [reply]
That shouldn't be needed. The operator application is using the wrong ceed

**Reply by NamjaeChoi** [reply]
I can't say anything with my knowledge. What I saw was that the parent of a fallback `Ceed` was the fallback itself, so I tried adding that line.

**Reply by jeremylt** [reply]
Now that I know the code is in your fork of the Moose repo, I can trace the issue there and see what's happening.

**Reply by jrwrigh** [reply]
> The code itself is already in my fork. It's just that compiling and running MOOSE will be a pain for you.

Just being able to read the code is quite helpful, even if we don't attempt to compile it.

Edit: Just saw that Jeremy said more-or-less the same thing.

**Reply by jeremylt** [reply]
Alright, give the updated branch a try

**Reply by NamjaeChoi** [reply]
Works now. The next issue is now on our side. Thanks for the quick fix!

**Reply by lindsayad** [reply]
The particular branch @NamjaeChoi is working in is https://github.com/NamjaeChoi/moose/tree/libceed

**Reply by lindsayad** [reply]
We're probably going to put our ceed exploration on the backburner for a bit as we haven't figured out a good way yet to incorporate it into MOOSE's object oriented design

**Reply by NamjaeChoi** [reply]
A potential way to realize complicated QFunctions could be leveraging JIT link. I created a small repository demonstrating the concept, which is based on the nvJitLink example (https://docs.nvidia.com/cuda/nvjitlink/index.html#example-device-lto-link-time-optimization).

https://github.com/NamjaeChoi/jit_test

In the example, a kernel `saxpy` is JIT-compiled, and it calls a device function `compute` which is statically compiled. `offline.cu` that contains the body of `compute` also contains some host codes, and it is compiled as a standard object file `offline.o`. It is then JIT-linked with the JIT-compiled kernel `saxpy` as well as statically linked to the executable. In this way, users can use QFunctions as simple entry points that merely call the statically-compiled user device functions which can have all the complicated things like our object-oriented design.

This will require libCEED to be able to receive the list of object files and do JIT-link. Also, a capability to receive QFunction source strings instead of source files in libCEED would be favorable, so that we don't have to create dummy source files that contain the QFunctions merely calling the user device functions.

This is just a quick thought that I wanted to share and not a request of any urgent implementation.

**Reply by jeremylt** [reply]
Hmm, how would that work without breaking the ability to select between CPU and GPU at runtime?

**Reply by NamjaeChoi** [reply]
Given the user provided proper function qualifiers (`__host__ __device__`) to his precompiled function, I wonder if there is a way for the CPU JIT compiler to link with the precompiled function.

**Reply by jeremylt** [reply]
CPU backends do not have JiT

**Reply by NamjaeChoi** [reply]
Then I guess it should work without any special treatment, because the object file is always statically linked. JIT link is an additional phase for GPU.
